File: sdks/hsafa-sdk/src/sdk.ts
// =============================================================================
// @hsafa/sdk — HsafaSDK class
// =============================================================================

import type {
  SdkOptions,
  ToolDefinition,
  ToolHandler,
  PushEventPayload,
  SdkEventType,
  SdkEventMap,
  ToolCallContext,
  Haseef,
  CreateHaseefInput,
  UpdateHaseefInput,
  SemanticMemory,
  SemanticMemoryInput,
  EpisodicMemory,
  SocialMemory,
  ProceduralMemory,
  MemoryStats,
  Run,
  ListRunsOptions,
} from './types.js';
import { inputToJsonSchema, parsePartialJson } from './schema.js';

const DEFAULT_RECONNECT_DELAY = 2_000;
const MAX_RECONNECT_DELAY = 30_000;

export class HsafaSDK {
  private readonly coreUrl: string;
  private readonly apiKey: string;
  readonly skill: string;

  private toolHandlers = new Map<string, ToolHandler>();
  private eventListeners = new Map<string, Set<(data: unknown) => void>>();
  private isConnected = false;
  private abortController: AbortController | null = null;

  constructor(opts: SdkOptions) {
    this.coreUrl = opts.coreUrl.replace(/\/$/, '');
    this.apiKey = opts.apiKey;
    this.skill = opts.skill;
  }

  // ── 1. REGISTER ─────────────────────────────────────────────────────────────

  async registerTools(tools: ToolDefinition[]): Promise<void> {
    const body = tools.map((t) => ({
      name: t.name,
      description: t.description,
      inputSchema: t.inputSchema ?? inputToJsonSchema(t.input ?? {}),
    }));

    const res = await fetch(`${this.coreUrl}/api/skills/${this.skill}/tools`, {
      method: 'PUT',
      headers: { 'x-api-key': this.apiKey, 'Content-Type': 'application/json' },
      body: JSON.stringify({ tools: body }),
    });

    if (!res.ok) {
      const text = await res.text().catch(() => '');
      throw new Error(`registerTools failed (${res.status}): ${text}`);
    }
  }

  // ── 2. HANDLE ────────────────────────────────────────────────────────────────

  onToolCall(name: string, handler: ToolHandler): void {
    this.toolHandlers.set(name, handler);
  }

  // ── 3. PUSH ──────────────────────────────────────────────────────────────────

  async pushEvent(event: PushEventPayload): Promise<void> {
    const res = await fetch(`${this.coreUrl}/api/events`, {
      method: 'POST',
      headers: { 'x-api-key': this.apiKey, 'Content-Type': 'application/json' },
      body: JSON.stringify({ skill: this.skill, ...event }),
    });

    if (!res.ok) {
      const text = await res.text().catch(() => '');
      throw new Error(`pushEvent failed (${res.status}): ${text}`);
    }
  }

  // ── Haseef API ───────────────────────────────────────────────────────────────

  readonly haseef = {
    list: async (): Promise<Haseef[]> => {
      const data = await this.request<{ haseefs: Haseef[] }>('GET', '/api/haseefs');
      return data.haseefs;
    },

    get: async (id: string): Promise<Haseef> => {
      const data = await this.request<{ haseef: Haseef }>('GET', `/api/haseefs/${id}`);
      return data.haseef;
    },

    create: async (input: CreateHaseefInput): Promise<Haseef> => {
      const data = await this.request<{ haseef: Haseef }>('POST', '/api/haseefs', input);
      return data.haseef;
    },

    update: async (id: string, patch: UpdateHaseefInput): Promise<Haseef> => {
      const data = await this.request<{ haseef: Haseef }>('PATCH', `/api/haseefs/${id}`, patch);
      return data.haseef;
    },

    delete: async (id: string): Promise<void> => {
      await this.request<unknown>('DELETE', `/api/haseefs/${id}`);
    },

    getProfile: async (id: string): Promise<Record<string, unknown>> => {
      const data = await this.request<{ profile: Record<string, unknown> }>('GET', `/api/haseefs/${id}/profile`);
      return data.profile ?? {};
    },

    updateProfile: async (id: string, patch: Record<string, unknown>): Promise<Record<string, unknown>> => {
      const data = await this.request<{ profile: Record<string, unknown> }>('PATCH', `/api/haseefs/${id}/profile`, patch);
      return data.profile ?? {};
    },

    addSkill: async (id: string, skillName: string): Promise<Haseef> => {
      const haseef = await this.haseef.get(id);
      const current = haseef.skills ?? [];
      if (current.includes(skillName)) return haseef;
      return this.haseef.update(id, { skills: [...current, skillName] });
    },

    removeSkill: async (id: string, skillName: string): Promise<Haseef> => {
      const haseef = await this.haseef.get(id);
      const current = haseef.skills ?? [];
      if (!current.includes(skillName)) return haseef;
      return this.haseef.update(id, { skills: current.filter((s) => s !== skillName) });
    },

    status: async (id: string): Promise<{ running: boolean; activeRunId: string | null }> => {
      return this.request('GET', `/api/haseefs/${id}/status`);
    },
  };

  // ── Memory API ───────────────────────────────────────────────────────────────

  readonly memory = {
    list: async (haseefId: string): Promise<SemanticMemory[]> => {
      const data = await this.request<{ memories: SemanticMemory[] }>('GET', `/api/memory/${haseefId}/semantic`);
      return data.memories;
    },

    search: async (haseefId: string, query: string, limit = 20): Promise<SemanticMemory[]> => {
      const qs = `?q=${encodeURIComponent(query)}&limit=${limit}`;
      const data = await this.request<{ results: SemanticMemory[] }>('GET', `/api/memory/${haseefId}/semantic/search${qs}`);
      return data.results;
    },

    set: async (haseefId: string, memories: SemanticMemoryInput[]): Promise<{ stored: number }> => {
      return this.request('POST', `/api/memory/${haseefId}/semantic`, { memories });
    },

    delete: async (haseefId: string, keys: string[]): Promise<{ deleted: number }> => {
      return this.request('DELETE', `/api/memory/${haseefId}/semantic`, { keys });
    },

    episodes: async (haseefId: string, limit = 20): Promise<EpisodicMemory[]> => {
      const data = await this.request<{ episodes: EpisodicMemory[] }>('GET', `/api/memory/${haseefId}/episodic?limit=${limit}`);
      return data.episodes;
    },

    searchEpisodes: async (haseefId: string, query: string, limit = 10): Promise<EpisodicMemory[]> => {
      const qs = `?q=${encodeURIComponent(query)}&limit=${limit}`;
      const data = await this.request<{ results: EpisodicMemory[] }>('GET', `/api/memory/${haseefId}/episodic/search${qs}`);
      return data.results;
    },

    social: async (haseefId: string): Promise<SocialMemory[]> => {
      const data = await this.request<{ people: SocialMemory[] }>('GET', `/api/memory/${haseefId}/social`);
      return data.people;
    },

    procedural: async (haseefId: string): Promise<ProceduralMemory[]> => {
      const data = await this.request<{ patterns: ProceduralMemory[] }>('GET', `/api/memory/${haseefId}/procedural`);
      return data.patterns;
    },

    stats: async (haseefId: string): Promise<MemoryStats> => {
      return this.request('GET', `/api/memory/${haseefId}/stats`);
    },
  };

  // ── Runs API ─────────────────────────────────────────────────────────────────

  readonly runs = {
    list: async (opts: ListRunsOptions = {}): Promise<Run[]> => {
      const params = new URLSearchParams();
      if (opts.haseefId) params.set('haseefId', opts.haseefId);
      if (opts.status) params.set('status', opts.status);
      if (opts.limit) params.set('limit', String(opts.limit));
      const qs = params.toString();
      const data = await this.request<{ runs: Run[] }>('GET', `/api/runs${qs ? `?${qs}` : ''}`);
      return data.runs;
    },

    get: async (runId: string): Promise<Run> => {
      const data = await this.request<{ run: Run }>('GET', `/api/runs/${runId}`);
      return data.run;
    },
  };

  // ── 4. LISTEN ────────────────────────────────────────────────────────────────

  on<K extends SdkEventType>(event: K, listener: (data: SdkEventMap[K]) => void): void {
    if (!this.eventListeners.has(event)) {
      this.eventListeners.set(event, new Set());
    }
    this.eventListeners.get(event)!.add(listener as (data: unknown) => void);
  }

  off<K extends SdkEventType>(event: K, listener: (data: SdkEventMap[K]) => void): void {
    this.eventListeners.get(event)?.delete(listener as (data: unknown) => void);
  }

  // ── CONNECT ──────────────────────────────────────────────────────────────────

  connect(): void {
    if (this.isConnected) return;
    this.isConnected = true;
    void this.sseLoop();
  }

  disconnect(): void {
    this.isConnected = false;
    this.abortController?.abort();
    this.abortController = null;
  }

  // ── Internals ────────────────────────────────────────────────────────────────

  private emit(event: string, data: unknown): void {
    const listeners = this.eventListeners.get(event);
    if (!listeners) return;
    for (const l of listeners) {
      try { l(data); } catch { /* swallow listener errors */ }
    }
  }

  private async sseLoop(): Promise<void> {
    let delay = DEFAULT_RECONNECT_DELAY;

    while (this.isConnected) {
      try {
        this.abortController = new AbortController();
        await this.openSSE(this.abortController.signal);
        delay = DEFAULT_RECONNECT_DELAY;
      } catch {
        if (!this.isConnected) break;
        await new Promise((r) => setTimeout(r, delay));
        delay = Math.min(delay * 2, MAX_RECONNECT_DELAY);
      }
    }
  }

  private async openSSE(signal: AbortSignal): Promise<void> {
    const url = `${this.coreUrl}/api/skills/${this.skill}/actions/stream`;
    const res = await fetch(url, {
      headers: { 'x-api-key': this.apiKey, Accept: 'text/event-stream' },
      signal,
    });

    if (!res.ok || !res.body) {
      throw new Error(`SSE connection failed: ${res.status}`);
    }

    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split('\n');
      buffer = lines.pop() ?? '';

      let dataLine = '';
      for (const line of lines) {
        if (line.startsWith('data: ')) {
          dataLine = line.slice(6).trim();
        } else if (line === '' && dataLine) {
          try {
            await this.handleMessage(JSON.parse(dataLine));
          } catch { /* ignore parse errors */ }
          dataLine = '';
        }
      }
    }
  }

  private async handleMessage(msg: Record<string, unknown>): Promise<void> {
    const type = msg.type as string;

    // Lifecycle events → forward to on() listeners
    const lifecycleEvents: SdkEventType[] = [
      'tool.input.start', 'tool.input.delta', 'tool.call',
      'tool.result', 'tool.error', 'run.started', 'run.completed',
    ];
    if (lifecycleEvents.includes(type as SdkEventType)) {
      this.emit(type, msg.data);
      return;
    }

    // Action request → route to onToolCall handler
    if (type === 'action') {
      const { actionId, toolName, args, haseef } = msg as {
        actionId: string;
        toolName: string;
        args: Record<string, unknown>;
        haseef: ToolCallContext['haseef'];
      };

      const handler = this.toolHandlers.get(toolName);
      if (!handler) {
        await this.postResult(actionId, { error: `No handler registered for tool "${toolName}"` });
        return;
      }

      try {
        const result = await handler(args ?? {}, { actionId, haseef });
        await this.postResult(actionId, result);
      } catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        await this.postResult(actionId, { error: message });
      }
    }

    // tool.input.delta with accumulated args for partial parsing
    if (type === 'tool.input.delta.raw') {
      const data = msg.data as { actionId: string; toolName: string; accumulatedText: string; haseef: unknown };
      const partialArgs = parsePartialJson(data.accumulatedText);
      this.emit('tool.input.delta', {
        actionId: data.actionId,
        toolName: data.toolName,
        delta: data.accumulatedText,
        partialArgs,
        haseef: data.haseef,
      });
    }
  }

  private async postResult(actionId: string, result: unknown): Promise<void> {
    try {
      await fetch(`${this.coreUrl}/api/actions/${actionId}/result`, {
        method: 'POST',
        headers: { 'x-api-key': this.apiKey, 'Content-Type': 'application/json' },
        body: JSON.stringify({ result }),
      });
    } catch (err) {
      console.error(`[HsafaSDK:${this.skill}] Failed to submit result for action ${actionId}:`, err);
    }
  }

  // Generic typed request helper for Core JSON endpoints.
  private async request<T>(
    method: 'GET' | 'POST' | 'PATCH' | 'DELETE',
    path: string,
    body?: unknown,
  ): Promise<T> {
    const init: RequestInit = {
      method,
      headers: { 'x-api-key': this.apiKey, 'Content-Type': 'application/json' },
    };
    if (body !== undefined) init.body = JSON.stringify(body);

    const res = await fetch(`${this.coreUrl}${path}`, init);
    if (!res.ok) {
      const text = await res.text().catch(() => '');
      throw new Error(`${method} ${path} failed (${res.status}): ${text}`);
    }
    if (res.status === 204) return undefined as T;
    const contentType = res.headers.get('content-type') ?? '';
    if (!contentType.includes('application/json')) return undefined as T;
    return (await res.json()) as T;
  }
}

File: sdks/python/src/hsafa_sdk/sdk.py
import asyncio
import inspect
import json
from typing import Dict, Any, List, Optional, Callable, Set
from urllib.parse import urlencode

from .types import (
    SdkOptions,
    ToolDefinition,
    ToolHandler,
    PushEventPayload,
    ToolCallContext,
    HaseefContext,
)
from .schema import input_to_json_schema, parse_partial_json

DEFAULT_RECONNECT_DELAY = 2.0
MAX_RECONNECT_DELAY = 30.0
DEFAULT_API_BASE = '/api/v7'


class HaseefAPI:
    def __init__(self, sdk: 'HsafaSDK'):
        self._sdk = sdk

    async def list(self) -> List[Dict[str, Any]]:
        res = await self._sdk._request("GET", f"{self._sdk.api_base}/haseefs")
        return res.get("haseefs", [])

    async def get(self, haseef_id: str) -> Dict[str, Any]:
        res = await self._sdk._request("GET", f"{self._sdk.api_base}/haseefs/{haseef_id}")
        return res.get("haseef", {})

    async def create(self, input_data: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._sdk._request("POST", f"{self._sdk.api_base}/haseefs", body=input_data)
        return res.get("haseef", {})

    async def update(self, haseef_id: str, patch: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._sdk._request("PATCH", f"{self._sdk.api_base}/haseefs/{haseef_id}", body=patch)
        return res.get("haseef", {})

    async def delete(self, haseef_id: str) -> None:
        await self._sdk._request("DELETE", f"{self._sdk.api_base}/haseefs/{haseef_id}")

    async def get_profile(self, haseef_id: str) -> Dict[str, Any]:
        res = await self._sdk._request("GET", f"{self._sdk.api_base}/haseefs/{haseef_id}/profile")
        return res.get("profile", {})

    async def update_profile(self, haseef_id: str, patch: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._sdk._request("PATCH", f"{self._sdk.api_base}/haseefs/{haseef_id}/profile", body=patch)
        return res.get("profile", {})

    async def add_skill(self, haseef_id: str, skill_name: str) -> Dict[str, Any]:
        haseef = await self.get(haseef_id)
        current = haseef.get("skills") or []
        if skill_name in current:
            return haseef
        return await self.update(haseef_id, {"skills": [*current, skill_name]})

    async def remove_skill(self, haseef_id: str, skill_name: str) -> Dict[str, Any]:
        haseef = await self.get(haseef_id)
        current = haseef.get("skills") or []
        if skill_name not in current:
            return haseef
        return await self.update(haseef_id, {"skills": [s for s in current if s != skill_name]})

    async def status(self, haseef_id: str) -> Dict[str, Any]:
        return await self._sdk._request("GET", f"{self._sdk.api_base}/haseefs/{haseef_id}/status")


class MemoryAPI:
    def __init__(self, sdk: 'HsafaSDK'):
        self._sdk = sdk

    async def list(self, haseef_id: str) -> List[Dict[str, Any]]:
        res = await self._sdk._request("GET", f"{self._sdk.api_base}/memory/{haseef_id}/semantic")
        return res.get("memories", [])

    async def search(self, haseef_id: str, query: str, limit: int = 20) -> List[Dict[str, Any]]:
        qs = urlencode({"q": query, "limit": limit})
        res = await self._sdk._request("GET", f"{self._sdk.api_base}/memory/{haseef_id}/semantic/search?{qs}")
        return res.get("results", [])

    async def set(self, haseef_id: str, memories: List[Dict[str, Any]]) -> Dict[str, Any]:
        return await self._sdk._request("POST", f"{self._sdk.api_base}/memory/{haseef_id}/semantic", body={"memories": memories})

    async def delete(self, haseef_id: str, keys: List[str]) -> Dict[str, Any]:
        return await self._sdk._request("DELETE", f"{self._sdk.api_base}/memory/{haseef_id}/semantic", body={"keys": keys})

    async def episodes(self, haseef_id: str, limit: int = 20) -> List[Dict[str, Any]]:
        res = await self._sdk._request("GET", f"{self._sdk.api_base}/memory/{haseef_id}/episodic?limit={limit}")
        return res.get("episodes", [])

    async def search_episodes(self, haseef_id: str, query: str, limit: int = 10) -> List[Dict[str, Any]]:
        qs = urlencode({"q": query, "limit": limit})
        res = await self._sdk._request("GET", f"{self._sdk.api_base}/memory/{haseef_id}/episodic/search?{qs}")
        return res.get("results", [])

    async def social(self, haseef_id: str) -> List[Dict[str, Any]]:
        res = await self._sdk._request("GET", f"{self._sdk.api_base}/memory/{haseef_id}/social")
        return res.get("people", [])

    async def procedural(self, haseef_id: str) -> List[Dict[str, Any]]:
        res = await self._sdk._request("GET", f"{self._sdk.api_base}/memory/{haseef_id}/procedural")
        return res.get("patterns", [])

    async def stats(self, haseef_id: str) -> Dict[str, Any]:
        return await self._sdk._request("GET", f"{self._sdk.api_base}/memory/{haseef_id}/stats")


class RunsAPI:
    def __init__(self, sdk: 'HsafaSDK'):
        self._sdk = sdk

    async def list(
        self,
        haseef_id: Optional[str] = None,
        status: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> List[Dict[str, Any]]:
        params: Dict[str, Any] = {}
        if haseef_id:
            params["haseefId"] = haseef_id
        if status:
            params["status"] = status
        if limit is not None:
            params["limit"] = limit
        qs = urlencode(params)
        path = f"{self._sdk.api_base}/runs" + (f"?{qs}" if qs else "")
        res = await self._sdk._request("GET", path)
        return res.get("runs", [])

    async def get(self, run_id: str) -> Dict[str, Any]:
        res = await self._sdk._request("GET", f"{self._sdk.api_base}/runs/{run_id}")
        return res.get("run", {})


class HsafaSDK:
    def __init__(self, opts: SdkOptions):
        self.core_url = opts.get("core_url", "http://localhost:3001").rstrip("/")
        self.api_key = opts.get("api_key", "")
        self.skill = opts.get("skill", "")
        self.api_base = (opts.get("api_base") or DEFAULT_API_BASE).rstrip("/")

        self._tool_handlers: Dict[str, ToolHandler] = {}
        self._event_listeners: Dict[str, Set[Callable[[Any], Any]]] = {}
        self._is_connected = False
        self._client = httpx.AsyncClient()

        self.haseef = HaseefAPI(self)
        self.memory = MemoryAPI(self)
        self.runs = RunsAPI(self)

    async def register_tools(self, tools: List[ToolDefinition]) -> None:
        body = []
        for t in tools:
            schema = t.get("inputSchema") or input_to_json_schema(t.get("input") or {})
            body.append({
                "name": t["name"],
                "description": t["description"],
                "inputSchema": schema,
            })

        path = f"{self.api_base}/skills/{self.skill}/tools"
        await self._request("PUT", path, body={"tools": body})

    def on_tool_call(self, name: str, handler: ToolHandler) -> None:
        self._tool_handlers[name] = handler

    async def push_event(self, event: PushEventPayload) -> None:
        payload = {"skill": self.skill, **event}
        await self._request("POST", f"{self.api_base}/events", body=payload)

    def on(self, event_name: str, listener: Callable[[Any], Any]) -> None:
        if event_name not in self._event_listeners:
            self._event_listeners[event_name] = set()
        self._event_listeners[event_name].add(listener)

    def off(self, event_name: str, listener: Callable[[Any], Any]) -> None:
        if event_name in self._event_listeners:
            self._event_listeners[event_name].discard(listener)

    async def _emit(self, event_name: str, data: Any) -> None:
        listeners = self._event_listeners.get(event_name, set())
        for listener in listeners:
            try:
                if inspect.iscoroutinefunction(listener):
                    await listener(data)
                else:
                    listener(data)
            except Exception as e:
                print(f"[HsafaSDK] Listener error on {event_name}: {e}")

    async def connect(self) -> None:
        """
        Connects to the SSE stream and blocks until disconnect() is called.
        Wrap in asyncio.create_task() to run in background alongside other code:
            asyncio.create_task(sdk.connect())
        """
        if self._is_connected:
            return
        self._is_connected = True

        delay = DEFAULT_RECONNECT_DELAY
        while self._is_connected:
            try:
                await self._open_sse()
                delay = DEFAULT_RECONNECT_DELAY
            except asyncio.CancelledError:
                break
            except Exception as e:
                if not self._is_connected:
                    break
                print(f"[HsafaSDK] SSE Connection lost ({e}). Reconnecting in {delay}s...")
                await asyncio.sleep(delay)
                delay = min(delay * 2, MAX_RECONNECT_DELAY)

    async def disconnect(self) -> None:
        self._is_connected = False
        await self._client.aclose()

    async def _open_sse(self) -> None:
        url = f"{self.core_url}{self.api_base}/skills/{self.skill}/actions/stream"
        headers = {
            "x-api-key": self.api_key,
            "Accept": "text/event-stream",
        }

        async with self._client.stream("GET", url, headers=headers, timeout=None) as response:
            response.raise_for_status()

            data_line = ""
            async for line in response.aiter_lines():
                if line.startswith("data: "):
                    data_line = line[6:].strip()
                elif line == "" and data_line:
                    try:
                        msg = json.loads(data_line)
                        asyncio.create_task(self._handle_message(msg))
                    except json.JSONDecodeError:
                        pass
                    data_line = ""

    async def _handle_message(self, msg: Dict[str, Any]) -> None:
        msg_type = msg.get("type")

        lifecycle_events = [
            "tool.input.start",
            "tool.input.delta",
            "tool.call",
            "tool.result",
            "tool.error",
            "run.started",
            "run.completed",
        ]

        if msg_type in lifecycle_events:
            await self._emit(msg_type, msg.get("data"))
            return

        if msg_type == "action":
            action_id = msg.get("actionId")
            tool_name = msg.get("toolName")
            args = msg.get("args", {})
            haseef = msg.get("haseef", {})

            handler = self._tool_handlers.get(tool_name)
            if not handler:
                await self._post_result(action_id, {"error": f'No handler registered for tool "{tool_name}"'})
                return

            try:
                ctx: ToolCallContext = {"actionId": action_id, "haseef": haseef}
                result = await handler(args, ctx)
                await self._post_result(action_id, result)
            except Exception as e:
                await self._post_result(action_id, {"error": str(e)})

        if msg_type == "tool.input.delta.raw":
            data = msg.get("data", {})
            partial_text = data.get("accumulatedText", "")
            await self._emit("tool.input.delta", {
                "actionId": data.get("actionId"),
                "toolName": data.get("toolName"),
                "delta": partial_text,
                "partialArgs": parse_partial_json(partial_text),
                "haseef": data.get("haseef"),
            })

    async def _post_result(self, action_id: str, result: Any) -> None:
        try:
            path = f"{self.api_base}/actions/{action_id}/result"
            await self._request("POST", path, body={"result": result})
        except Exception as e:
            print(f"[HsafaSDK:{self.skill}] Failed to submit result for action {action_id}: {e}")

    async def _request(
        self,
        method: str,
        path: str,
        body: Optional[Dict[str, Any]] = None,
    ) -> Any:
        url = f"{self.core_url}{path}"
        headers = {
            "x-api-key": self.api_key,
            "Content-Type": "application/json",
        }

        response = await self._client.request(method, url, headers=headers, json=body)

        if not response.is_success:
            raise Exception(f"{method} {path} failed ({response.status_code}): {response.text}")

        if response.status_code == 204 or not response.content:
            return None

        if "application/json" in response.headers.get("content-type", ""):
            return response.json()

        return None


import httpx
